File: data/code_pairs/cpp/2508.cpp
#include <iostream>
#include <cmath>
#include <vector>
#include <cassert>

int IDAMAX(int N, const std::vector<double>& DX, int INCX) {
    if (N <= 0) return 0;
    int IDAMAX = 1;
    if (N == 1) return IDAMAX;
    double DMAX = std::abs(DX[0]);
    int IX = 0;
    
    if (INCX != 1) {
        if (INCX < 0) IX = (-N + 1) * INCX;
        DMAX = std::abs(DX[IX]);
        IX += INCX;
        for (int I = 2; I <= N; ++I) {
            double XMAG = std::abs(DX[IX]);
            if (XMAG > DMAX) {
                IDAMAX = I;
                DMAX = XMAG;
            }
            IX += INCX;
        }
    } else {
        for (int I = 1; I < N; ++I) {
            double XMAG = std::abs(DX[I]);
            if (XMAG > DMAX) {
                IDAMAX = I + 1;
                DMAX = XMAG;
            }
        }
    }
    return IDAMAX;
}

int main() {
    std::vector<double> dx = {1.0, 3.0, -5.0, 2.0, 4.0};
    int n = dx.size();
    int incx = 1;
    
    // Expected result: 3 (since | -5.0 | is the largest)
    int expected_result = 3;

    // Call the IDAMAX function
    int result = IDAMAX(n, dx, incx);

    // Print the result
    std::cout << "IDAMAX result: " << result << std::endl;

    // Verify the result
    assert(result == expected_result);
    
    std::cout << "Unit test passed!" << std::endl;

    return 0;
}

File: data/code_pairs/fortran/2508.f90
program test_idamax
  implicit none
  integer :: n, incx, result
  double precision, allocatable :: dx(:)
  
  ! Initialize test data
  n = 5
  incx = 1
  allocate(dx(n))
  dx = (/1.0d0, 3.0d0, -5.0d0, 2.0d0, 4.0d0/)
  
  ! Call the IDAMAX function
  result = idamax(n, dx, incx)
  
  ! Print the result
  print *, 'IDAMAX result:', result
  
contains

  INTEGER FUNCTION IDAMAX(N, DX, INCX)
    INTEGER, INTENT(IN) :: N, INCX
    DOUBLE PRECISION, INTENT(IN) :: DX(*)
    DOUBLE PRECISION :: DMAX, XMAG
    INTEGER :: I, IX
    
    IDAMAX = 0
    IF (N.LE.0) RETURN
    IDAMAX = 1
    IF (N.EQ.1) RETURN
    IF (INCX.EQ.1) GOTO 20
    IX = 1
    IF (INCX.LT.0) IX = (-N+1)*INCX + 1
    DMAX = ABS(DX(IX))
    IX = IX + INCX
    DO I = 2,N
      XMAG = ABS(DX(IX))
      IF (XMAG.GT.DMAX) THEN
        IDAMAX = I
        DMAX = XMAG
      END IF
      IX = IX + INCX
    END DO
    RETURN
20  DMAX = ABS(DX(1))
    DO I = 2,N
      XMAG = ABS(DX(I))
      IF (XMAG.GT.DMAX) THEN
        IDAMAX = I
        DMAX = XMAG
      END IF
    END DO
    RETURN
  END FUNCTION IDAMAX

end program test_idamax